File: Binary Tree/572. Subtree of Another Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sametree(self, one, two):
        if not one and not two:
            return True
        if one and two and one.val == two.val:
            return self.sametree(one.left,two.left) and self.sametree(one.right,two.right)
        return False

    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root: # If reached end of the root, it's false 
            return False
        if not subRoot: # Checking for default empty subRoots
            return True

        if self.sametree(root, subRoot):
            return True
        else:
            return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

File: Binary Tree/572. Subtree of Another Tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

bool isSame(struct TreeNode* root1, struct TreeNode* root2) {
    if (root1 == NULL && root2 == NULL){
        return true;
    }
    if (root1 && root2 && root1->val == root2->val){
        return isSame(root1->left, root2->left) && isSame(root1->right, root2->right);
    }
    return false; 
}

bool isSubtree(struct TreeNode* root, struct TreeNode* subRoot){
    if (root==NULL) {
        return false;
    }
    if (subRoot == NULL) {
        return true;
    }

    if (isSame(root, subRoot)){
        return true;
    }
    else {
        return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
    }
}
